crates/services/executor/src/executor.rs

```diff
@@ -624,7 +659,144 @@ where
         }
 
         data.changes = self.block_st_transaction.into_changes();
+        Ok(data)
+    }
+
+    // TODO: Refactor this function to reduce the number of arguments.
+    #[allow(clippy::too_many_arguments)]
+    fn execute_transaction_and_commit<
+        W: WriteTransaction + KeyValueInspect<Column = Column> + Modifiable,
+    >(
+        &self,
+        block: &mut PartialFuelBlock,
+        thread_block_transaction: &mut W,
+        execution_data: &mut ExecutionData,
+        tx: MaybeCheckedTransaction,
+        gas_price: Word,
+        coinbase_contract_id: ContractId,
+        execution_kind: ExecutionKind,
+    ) -> ExecutorResult<()> {
+        let tx_count = execution_data.tx_count;
+        let tx = {
+            let mut tx_st_transaction = thread_block_transaction
+                .write_transaction()
+                .with_policy(ConflictPolicy::Overwrite);
+            let tx_id = tx.id(&self.consensus_params.chain_id());
+            let tx = self.execute_transaction(
+                tx,
+                &tx_id,
+                &block.header,
+                coinbase_contract_id,
+                gas_price,
+                execution_data,
+                execution_kind,
+                &mut tx_st_transaction,
+            )?;
+            tx_st_transaction.commit()?;
+            tx
+        };
+
+        block.transactions.push(tx);
+        execution_data.tx_count = tx_count
+            .checked_add(1)
+            .ok_or(ExecutorError::TooManyTransactions)?;
+
+        Ok(())
+    }
 
+    #[tracing::instrument(skip_all)]
+    fn validate_block(
+        mut self,
+        component: PartialBlockComponent<OnceTransactionsSource>,
+    ) -> ExecutorResult<ExecutionData> {
+        let block_gas_limit = self.consensus_params.block_gas_limit();
+        let mut data = ExecutionData::new();
+
+        let block = component.empty_block;
+        let source = component.transactions_source;
+        let gas_price = component.gas_price;
+        let coinbase_contract_id = component.coinbase_contract_id;
+        let block_height = *block.header.height();
+        let execution_kind = ExecutionKind::Validation;
+
+        let forced_transactions = if self.relayer.enabled() {
```

Comment:
Maybe it make sense to move relayer related logic into a separate functions and reuse it in `validate_block` and `execute_block`=) It can be a part of the follow up